sdk/keyvault/Azure.Security.KeyVault.Administration/src/KeyVaultAccessControlClient.cs

```diff
@@ -151,11 +151,138 @@ public virtual AsyncPageable<KeyVaultRoleDefinition> GetRoleDefinitionsAsync(Key
             });
         }
 
+        /// <summary>
+        /// Creates or updates a role definition.
+        /// </summary>
+        /// <param name="roleDefinitionDescription">The description for the role definition.</param>
+        /// <param name="permissions">The permissions granted by the role definition when assigned to a principal.</param>
+        /// <param name="roleScope">The scope of the <see cref="KeyVaultRoleDefinition"/> to create. The default value is <see cref="KeyVaultRoleScope.Global"/>.</param>
+        /// <param name="roleDefinitionName">Optional name used to create the role definition. A new <see cref="Guid"/> will be generated if not specified.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
+        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
+        public virtual async Task<Response<KeyVaultRoleDefinition>> UpsertRoleDefinitionAsync(string roleDefinitionDescription, KeyVaultPermission permissions, KeyVaultRoleScope roleScope = default, Guid? roleDefinitionName = null, CancellationToken cancellationToken = default)
+        {
+            using DiagnosticScope scope = _diagnostics.CreateScope($"{nameof(KeyVaultAccessControlClient)}.{nameof(UpsertRoleDefinition)}");
+            scope.Start();
+            try
+            {
+                var name = (roleDefinitionName ?? Guid.NewGuid()).ToString();
+                var properties = new RoleDefinitionProperties()
+                {
+                    Description = roleDefinitionDescription,
+                    RoleName = name,
+                    RoleType = KeyVaultRoleType.CustomRole
+                };
+                properties.AssignableScopes.Add(roleScope);
+                properties.Permissions.Add(permissions);
+
+                var parameters = new RoleDefinitionCreateParameters(properties);
+
+                return await _definitionsRestClient.CreateOrUpdateAsync(
```

Comment:
Given the generator is using `CreateOrUpdate`, I'd throw my $0.02 USD behind that name.

Reply:
We used Upsert in tables and it sounds more succinct to me. Happy to go with the majority / guidelines though.

Reply:
`CreateOrUpdate` is far more prevalent in track 2 sources in this repo:

```powershell
$matches = dir sdk\* | dir -include Azure.* -exclude Azure.ResourceManager.*, Azure.Management.* -dir `
  | join-path -child src -resolve | dir -filter *.cs -recurse | select-string 'Upsert|CreateOrUpdate' `
  | group @{e={$_.Matches[0].Value}} -noelem
```
```text
Count Name
----- ----
  215 CreateOrUpdate
   42 upsert
```

Reply:
`Upsert` tends to mean different things to different folks, unfortunately.  At this point I learn toward `CreateOrReplace`/`CreateOrUpdate` and `CreateOrMerge` as two separate names.

Reply:
I have an internal thread on this and sounds like we're converging toward:

* Set: creates or replaces a resource entirely.
* Upsert: creates or updates a resources in a database.
* CreateOrUpdate: creates or updates a resource in something other than a database.

Reply:
"Set" probably isn't good in most cases. It seems to work for AppConfig values or KV secrets where the whole things is replaced (ignoring metadata), but seldom else.